feat: view shared pipes

kevinkim-ogp opened the pull request:
feat: view shared pipes

chore: remove debug statement

chore: add role to getFlows

chore: allow collaborators to view pipe details

kevinkim-ogp commented:
* **#1283** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1283?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1280** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1280?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1278** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1278?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1277** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1277?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1252** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1252?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1245** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1245?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1220** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1220?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1196** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1196?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1195** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1195?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1256** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1256?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1194** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1194?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1193** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1193?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1153** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1152** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1151** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1151?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1150** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1150?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1191** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1191?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1190** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1190?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1149** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1149?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1149?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1146** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1146?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#1175](https://github.com/opengovsg/plumber/pull/1175) <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1175?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#1174** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1174?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1145** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1145?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1173** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1173?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1144** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1144?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `trunk/collaborators`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

pregnantboy commented on packages/backend/src/models/user.ts:
```diff
@@ -148,40 +148,136 @@ class User extends Base {
+    trx?: Transaction
+    requiredRole: IFlowCollabRole
+  }): ExtendedQueryBuilder<Step, Step[]>
+  withAccessible({
```
since the joins are quite different, how bout simplying to withAccessibleSteps, withAccessibleFlows etc...

pregnantboy replied:
probably need to remove the existing `relationMappings` for conenctions, flows, steps, executions to prevent accidentally using them

pregnantboy commented on packages/backend/src/apps/m365-excel/common/file-privacy.ts:
```diff
@@ -209,6 +209,16 @@ export async function validateCanAccessFile(
+   *
+   * the actual Pipe execution is still in the context of the Pipe owner.
+   */
+  collaborators: IFlowCollaborator[],
```
it seems that validateCanAccessFile is mainly used for 2 things: 
1. dynamic data. 
2. acquiring sessions for non-test runs

how bout just check for permission in the get-dynamic-data handler and use the flow's owner for userId?

pregnantboy commented on packages/backend/src/graphql/queries/test-connection.ts:
```diff
@@ -10,20 +11,33 @@ const testConnection: QueryResolvers['testConnection'] = async (
 ) => {
   let connection = await context.currentUser
-    .$relatedQuery('connections')
+    .withAccessible({
+      type: 'connection',
+      requiredRole: 'viewer',
+    })
```
somehow this is preventing me from adding a new excel connection if it doesnt exists

pregnantboy commented on packages/backend/src/models/user.ts:
```diff
@@ -148,40 +148,136 @@ class User extends Base {
+          (queryBuilder as ExtendedQueryBuilder<Connection, Connection[]>) ||
+          Connection.query(trx)
+        baseQuery
+          .select(
+            'connections.*',
+            Connection.raw(USER_ROLE_STMT, [userId, userId]),
+          )
+          .join('steps', 'steps.connection_id', 'connections.id')
+          .join('flows', 'steps.flow_id', 'flows.id')
+          .leftJoin('flow_connections', 'flow_connections.flow_id', 'flows.id')
```
is there a need to join? can be it fetched from user's connections as well as flow_connections

pregnantboy commented on packages/backend/src/graphql/queries/get-flows.ts:
```diff
@@ -49,14 +56,20 @@ const getFlows: QueryResolvers['getFlows'] = async (
+
+  flowsQuery
     .withGraphFetched({
       steps: {
         connection: true,
       },
       pendingTransfer: true,
+      collaborators: {
+        user: true,
+      },
```
need to check which of these are still needed

pregnantboy commented on packages/backend/src/graphql/queries/get-flows.ts:
```diff
@@ -49,14 +56,20 @@ const getFlows: QueryResolvers['getFlows'] = async (
+    .innerJoin('filtered_steps', 'flows.id', 'filtered_steps.flow_id')
+    .whereIn('flows.id', filteredFlowIds)
+
+  flowsQuery
```
why is this split up and duplicated .whereIn('flow.id', filteredFlowIds) shouldn't be required?

pregnantboy commented on packages/backend/src/graphql/custom-resolvers/flow-collaborator.ts:
```diff
@@ -0,0 +1,21 @@
+    return parent?.user?.email
+  }
+
+  // edge case: fallback to query if user relation is not loaded
```
add a comment on where this is not alr loaded

pregnantboy commented on packages/backend/src/models/user.ts:
```diff
@@ -148,40 +148,136 @@ class User extends Base {
+    requiredRole: IFlowCollabRole,
+  ) {
+    if (requiredRole === 'owner') {
+      query.where('flows.user_id', userId)
+    } else if (requiredRole === 'editor') {
+      query.where(function () {
+        this.where('flows.user_id', userId).orWhereExists(function () {
+          this.select('*')
+            .from('flow_collaborators')
+            .whereRaw('flow_collaborators.flow_id = flows.id')
+            .where('flow_collaborators.user_id', userId)
+            .where('flow_collaborators.role', 'editor')
+            .whereNull('flow_collaborators.deleted_at')
+        })
+      })
+    } else if (requiredRole === 'viewer') {
+      query.where(function () {
+        this.where('flows.user_id', userId).orWhereExists(function () {
+          this.select('*')
+            .from('flow_collaborators')
+            .whereRaw('flow_collaborators.flow_id = flows.id')
+            .where('flow_collaborators.user_id', userId)
+            .whereIn('flow_collaborators.role', ['editor', 'viewer'])
+            .whereNull('flow_collaborators.deleted_at')
+        })
+      })
```
this can be simplified.

.where('flows.user_id', userId) is always present

.whereIn('...', ['editor', 'viewer']) <--- only this array changes if not owner perms

probably can see how to re-use the existing `checkUserPermission` helper file by adding a method to getAllowedRoles(requiredRole) => ['viewer', 'editor']

kevinkim-ogp commented:
changes made based on comments:
* refactored to separate withAccessible queries
* added `supportsConnectionRegistration` to the `testConnection` query to identify whether it needs to verify connection registration. it was previously determined by whether `flowId` was provided, but this no longer works as we need the `flowId` to check for shared connections in `flow_connections`

additional note:
* relational mappings will be removed at the end in the refactor PR
